### ISSUE TITLE
Can we have export/import history option

### ISSUE THREAD

**Issue by vvsudhir** [a85/POSTMan-Chrome-Extension#1647]
It would be helpful if we have similar functionality of export/import data for history in postman

**Comment by a85** [comment]
@vvsudhir Can you tell me more about the use case?

**Comment by vvsudhir** [comment]
suppose if I am using chrome extension of postman and I would like to move to app then I don't get data what I have in history section . If we have option for export/import history then I can get that data and also I can share history.

**Comment by a85** [comment]
For moving to the app you can use the Settings > Data > Export option and import it all into Postman app's history. For sharing requests, we recommend using the Collections feature. The Chrome extension is deprecated. I recommend using the Postman app + Interceptor on Windows. If you are on Mac, you can get our Mac app here that has in-built support for cookies: www.getpostman.com/apps

**Comment by vvsudhir** [comment]
I think "export data" option doesn't export data what we have in history section

**Comment by vvsudhir** [comment]
Sorry for the confusion, we have this option now.

**Comment by blaxter** [comment]
Actually the export data option doesn't export history, it only exports collections, environments, globals and header presets (and sync option is not working for me chrome app <-> desktop app).

**Comment by stephenmcd1** [comment]
I'm in the process of moving from the Chrome App version of Postman to the Native version and was also looking for a way to export/import my history.  I couldn't find anyway through the front-end, but I hacked together some javascript to do it.  It's not pretty but it worked!  There is probably a cleaner way to write the code (I was too lazy to use actual callbacks/event handlers so you can't execute all the code at once).  Here is a dump of my solution...if anybody actually tries to use it and runs into problems, let me know and I can elaborate on any confusing parts:

```javascript
//In Chrome DevTools on the background page of the Postman extension...

//A handy helper method that lets you save data from the console to a file
(function(console){

    console.save = function(data, filename){

        if(!data) {
            console.error('Console.save: No data')
            return;
        }

        if(!filename) filename = 'console.json'

        if(typeof data === "object"){
            data = JSON.stringify(data, undefined, 4)
        }

        var blob = new Blob([data], {type: 'text/json'}),
            e    = document.createEvent('MouseEvents'),
            a    = document.createElement('a')

        a.download = filename
        a.href = window.URL.createObjectURL(blob)
        a.dataset.downloadurl =  ['text/json', a.download, a.href].join(':')
        e.initMouseEvent('click', true, false, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null)
        a.dispatchEvent(e)
    }
})(console)


var req = indexedDB.open('postman')

//Wait for that to finish

var db = req.result
var r = db.transaction(["requests"],"readwrite").objectStore('requests').getAll()

//Wait for that to finish

//This will download a text file with your history
console.save(JSON.stringify(r.result), 'postman-requests-export.json')


//Switch to standalone app and open the dev console

var req = indexedDB.open('postman')

//Wait for that to finish

var db = req.result

var data = //Paste the text from the file exported above

data.forEach(function(a){db.transaction(["requests"],"readwrite").objectStore('requests').add(a)})

//Restart Postman
```

**Comment by AgentMC** [comment]
@stephenmcd1 worked great, just couple of notes for the other guys who will use it:

1. This script must be executed from comment to comment. Read text in comments. (it is not obvious that those are meaningful until you read them)
2. I don't know for Chrome, but in Vivaldi, the file is saved straight in Downloads directory without any notifications at all.
3. To open "Chrome DevTools on the background page of the Postman extension", go to Menu > (More) Tools > Extensions > enable Developer mode > click "inspect Background Page" against Postman extension.

**Comment by ratatatKE** [comment]
This is perfect, worked like magic. as @AgentMC  said, make sure you execute the code one comment to the next.. to the next .. while reading the comments and replacing the `data` variable.

**Comment by barunhflipkart** [comment]
It is showing 'undefined' after running each statement. @stephenmcd1  @AgentMC

**Comment by AgentMC** [comment]
@barunhflipkart That is expected. All the output from each set of statements is stored in variables, so there is nothing to output to the console.

**Comment by barunhflipkart** [comment]
Yes, it has worked, thanks a lot.

**Comment by drice** [comment]
Code seemed to work for me. If I run 

```var req = indexedDB.open('postman')

//Wait for that to finish

var db = req.result
var r = db.transaction(["requests"],"readwrite").objectStore('requests').getAll()
```

in the App, I  see the files from the Web App, but I don't see it being populated in the History section. Any ideas?

**Comment by itsforgit** [comment]
**Thanks a lot** stephenmcd1. Its worked for me.. :)

**Comment by akgupta0304** [comment]
Best way which I used, Create a login on postman and whenever You do login on another computer login with the same account, You will get all the data.

Otherwise, Export the data into Json format from setting->data. And the same json can be used for import.

**Comment by ingenuity2k** [comment]
@stephenmcd1 thanks for your script. It works until I try to import the data, which is where I get a DOM Exception:
`Uncaught DOMException: Failed to execute 'transaction' on 'IDBDatabase': One of the specified object stores was not found.
    at <anonymous>:1:29
    at Array.forEach (native)
    at <anonymous>:1:6`

Not sure why it doesn't work. Perhaps they changed something db wise?

**Comment by smendez-hi** [comment]
@ingenuity2k I'm having the same issue. Any updates?

In the Webapp, `db.objectStoreNames` returns a lot of names, but in the native app the property is empty, so that's why the object store specified (requests) is not found. But i don't know how to create it.

EDIT: There are 2 databases in the native app (`indexedDB.webkitGetDatabaseNames()`), "postman" and "postman-app". The latter has all the objectStoreNames that the former has in the webapp. And the `forEach()` command of the script works. But restarting Postman does nothing. I tried it twice with no result.

**Comment by ingenuity2k** [comment]
@smendez-hi nope, no further progress from my side. 

I hope someone with more knowledge of the postman db/tables could provide some insight so we could modify @stephenmcd1 's script to adapt it to the apparent changes.

**Comment by stephenmcd1** [comment]
@smendez-hi, @ingenuity2k I'll try to take a look at this today to see if I can figure out why it's breaking....

**Comment by stephenmcd1** [comment]
It looks like things have moved around and changed a bit in the latest version.  As @smendez-hi noticed, there is the postman-app DB that seems to be where we need to import our data.  We also need to go to the history store instead of requests and also give each request the workspace it belongs to.  When I did those changes, my saved requests started showing up.  

And since more people have been using my script that I imagined, I spent the extra few minutes and wrote the code so you don't need to manually run fragments.

Here is a better version of the logic to export the requests to a text file.  None of the logic changed here, it's just all in one block of code that can be executed together:

```javascript
//A handy helper method that lets you save data from the console to a file
(function(console){

    console.save = function(data, filename){

        if(!data) {
            console.error('Console.save: No data')
            return;
        }

        if(!filename) filename = 'console.json'

        if(typeof data === "object"){
            data = JSON.stringify(data, undefined, 4)
        }

        var blob = new Blob([data], {type: 'text/json'}),
            e    = document.createEvent('MouseEvents'),
            a    = document.createElement('a')

        a.download = filename
        a.href = window.URL.createObjectURL(blob)
        a.dataset.downloadurl =  ['text/json', a.download, a.href].join(':')
        e.initMouseEvent('click', true, false, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null)
        a.dispatchEvent(e)
    }
})(console)

//Common error reporting function
function reportError(){
	console.error('Oops, something went wrong :-(');
}

//Open the database
var dbReq = indexedDB.open('postman')
dbReq.onerror = reportError;
dbReq.onsuccess = function(){
	var db = dbReq.result;
	
	//Query for all the saved requests
	var requestReq = db.transaction(["requests"],"readwrite").objectStore('requests').getAll();
	requestReq.onerror = reportError;
	requestReq.onsuccess = function(){
		var requests = requestReq.result;
		
		//Dump them to a file
		console.save(JSON.stringify(requests), 'postman-requests-export.json')
		console.info('Your existing requests have been exported to a file and downloaded to your computer.  You will need to copy the contents of that file for the next part')
	};
};
```
Then this is the updated code that you run in the dev console of the new postman app.  Notice there there is now two things you need to do at the beginning of the script:

```javascript
//Paste the text from the exported file here (overwriting the empty array)
var data = [] 

//Enter the guid/id of the workspace to import into.  Run the script with this value blank if you need some help
//  finding this value.  Also, be sure you don't end up with extra quotes if you copy/paste the value
var ws = ''; 

//Common error reporting function
function reportError(){
	console.error('Oops, something went wrong :-(');
}

//Open the database
var dbReq = indexedDB.open('postman-app')
dbReq.onerror = reportError;
dbReq.onsuccess = function(){
	var db = dbReq.result;
	
	if(!data.length){
		console.error('You did not pass in any exported requests so there is nothing for this script to do.  Perhaps you forgot to paste your request data?');
		return;
	}
	
	if(!ws){
		var wsReq = db.transaction(["workspace"],"readwrite").objectStore('workspace').getAll();
		wsReq.onerror = reportError;
		wsReq.onsuccess = function(){
			console.error('You did not specify a workspace.  Below is a dump of all your workspaces.  Grab the guid (ID field) from the workspace you want these requests to show up under and include it at the top of this script');
			console.log(wsReq.result);
		}
		return;
	}
	
	data.forEach(function(a){
		a.workspace = ws;
		db.transaction(["history"],"readwrite").objectStore('history').add(a);
	});
	
	console.log('Requests have been imported.  Give it a second to finish up and then restart Postman')
}
```
Give that new code a try and let me know how it goes.  Good luck!

**Comment by smendez-hi** [comment]
@stephenmcd1 It is working now! Thank you for your time fixing this.
One note: all the history will appear below "Today" as if every call has been made today. I don't care about this, anyway.

**Comment by ingenuity2k** [comment]
Sorry for the late reply. Just tried and it works like a charm. Thanks for taking the time to help out @stephenmcd1 !

**Comment by jrorke** [comment]
Any way to archive the history or console output from the standalone Postman app?  My company seems to block access to login to the Postman cloud for sync'ing.

**Comment by a85** [comment]
There are no plans for history exporting right now or in the near future. We recommend importing/exporting more specifically through collections.

**Comment by jamarciano** [comment]
Thank you for noting the issue with the dates @smendez-hi, because I do care a bit about such things. By debugging the script @stephenmcd1 so wonderfully provided (thank you so much for that!), I was able see that the old history used a "timestamp" field in milliseconds while the current history uses a "createdAt" Date. This very slightly updated version of stephen's import script will handle that too:

```js
//Paste the text from the exported file here (overwriting the empty array)
var data = []

//Enter the guid/id of the workspace to import into.  Run the script with this value blank if you need some help
//  finding this value.  Also, be sure you don't end up with extra quotes if you copy/paste the value
var ws = ''; 

//Common error reporting function
function reportError(){
	console.error('Oops, something went wrong :-(');
}

//Open the database
var dbReq = indexedDB.open('postman-app')
dbReq.onerror = reportError;
dbReq.onsuccess = function(){
	var db = dbReq.result;
	
	if(!data.length){
		console.error('You did not pass in any exported requests so there is nothing for this script to do.  Perhaps you forgot to paste your request data?');
		return;
	}
	
	if(!ws){
		var wsReq = db.transaction(["workspace"],"readwrite").objectStore('workspace').getAll();
		wsReq.onerror = reportError;
		wsReq.onsuccess = function(){
			console.error('You did not specify a workspace.  Below is a dump of all your workspaces.  Grab the guid (ID field) from the workspace you want these requests to show up under and include it at the top of this script');
			console.log(wsReq.result);
		}
		return;
	}
	
	data.forEach(function(a){
		a.workspace = ws;
		a.createdAt = new Date(a.timestamp);

		db.transaction(["history"],"readwrite").objectStore('history').add(a);
	});
	
	console.log('Requests have been imported.  Give it a second to finish up and then restart Postman')
}
```

**Comment by dcarbone** [comment]
If you're on a mac and want to be REALLY lazy about it, you can also just copy over the entirety of the `/{user}/Library/Application\ Support/Postman/*` to your new machine.  Seems to work.  Unintended side effects unknown.

**Comment by kirandesai** [comment]
> It looks like things have moved around and changed a bit in the latest version. As @smendez-hi noticed, there is the postman-app DB that seems to be where we need to import our data. We also need to go to the history store instead of requests and also give each request the workspace it belongs to. When I did those changes, my saved requests started showing up.
> 
> And since more people have been using my script that I imagined, I spent the extra few minutes and wrote the code so you don't need to manually run fragments.
> 
> Here is a better version of the logic to export the requests to a text file. None of the logic changed here, it's just all in one block of code that can be executed together:
> 
> ```js
> //A handy helper method that lets you save data from the console to a file
> (function(console){
> 
>     console.save = function(data, filename){
> 
>         if(!data) {
>             console.error('Console.save: No data')
>             return;
>         }
> 
>         if(!filename) filename = 'console.json'
> 
>         if(typeof data === "object"){
>             data = JSON.stringify(data, undefined, 4)
>         }
> 
>         var blob = new Blob([data], {type: 'text/json'}),
>             e    = document.createEvent('MouseEvents'),
>             a    = document.createElement('a')
> 
>         a.download = filename
>         a.href = window.URL.createObjectURL(blob)
>         a.dataset.downloadurl =  ['text/json', a.download, a.href].join(':')
>         e.initMouseEvent('click', true, false, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null)
>         a.dispatchEvent(e)
>     }
> })(console)
> 
> //Common error reporting function
> function reportError(){
> 	console.error('Oops, something went wrong :-(');
> }
> 
> //Open the database
> var dbReq = indexedDB.open('postman')
> dbReq.onerror = reportError;
> dbReq.onsuccess = function(){
> 	var db = dbReq.result;
> 	
> 	//Query for all the saved requests
> 	var requestReq = db.transaction(["requests"],"readwrite").objectStore('requests').getAll();
> 	requestReq.onerror = reportError;
> 	requestReq.onsuccess = function(){
> 		var requests = requestReq.result;
> 		
> 		//Dump them to a file
> 		console.save(JSON.stringify(requests), 'postman-requests-export.json')
> 		console.info('Your existing requests have been exported to a file and downloaded to your computer.  You will need to copy the contents of that file for the next part')
> 	};
> };
> ```
> 
> Then this is the updated code that you run in the dev console of the new postman app. Notice there there is now two things you need to do at the beginning of the script:
> 
> ```js
> //Paste the text from the exported file here (overwriting the empty array)
> var data = [] 
> 
> //Enter the guid/id of the workspace to import into.  Run the script with this value blank if you need some help
> //  finding this value.  Also, be sure you don't end up with extra quotes if you copy/paste the value
> var ws = ''; 
> 
> //Common error reporting function
> function reportError(){
> 	console.error('Oops, something went wrong :-(');
> }
> 
> //Open the database
> var dbReq = indexedDB.open('postman-app')
> dbReq.onerror = reportError;
> dbReq.onsuccess = function(){
> 	var db = dbReq.result;
> 	
> 	if(!data.length){
> 		console.error('You did not pass in any exported requests so there is nothing for this script to do.  Perhaps you forgot to paste your request data?');
> 		return;
> 	}
> 	
> 	if(!ws){
> 		var wsReq = db.transaction(["workspace"],"readwrite").objectStore('workspace').getAll();
> 		wsReq.onerror = reportError;
> 		wsReq.onsuccess = function(){
> 			console.error('You did not specify a workspace.  Below is a dump of all your workspaces.  Grab the guid (ID field) from the workspace you want these requests to show up under and include it at the top of this script');
> 			console.log(wsReq.result);
> 		}
> 		return;
> 	}
> 	
> 	data.forEach(function(a){
> 		a.workspace = ws;
> 		db.transaction(["history"],"readwrite").objectStore('history').add(a);
> 	});
> 	
> 	console.log('Requests have been imported.  Give it a second to finish up and then restart Postman')
> }
> ```
> 
> Give that new code a try and let me know how it goes. Good luck!

Thank you God.

**Comment by kirandesai** [comment]
> Give that new code a try and let me know how it goes. Good luck!


Thank you @smendez-hi @stephenmcd1 and @jamarciano. You guys are life savers.

**Comment by LuisTAC** [comment]
Thanks a lot @stephenmcd1 @jamarciano! I needed to search the request history for responses that included a specific string (which, sadly, postman doesn't yet support). By adapting your scripts, I managed to do just that without much previous knowledge in JS and none in the IndexedDb API!
Here's my final script if anyone needs it.
```javascript
//A handy helper method that lets you save data from the console to a file
(function(console){

    console.save = function(data, filename){

        if(!data) {
            console.error('Console.save: No data')
            return;
        }

        if(!filename) filename = 'console.json'

        if(typeof data === "object"){
            data = JSON.stringify(data, undefined, 4)
        }

        var blob = new Blob([data], {type: 'text/json'}),
            e    = document.createEvent('MouseEvents'),
            a    = document.createElement('a')

        a.download = filename
        a.href = window.URL.createObjectURL(blob)
        a.dataset.downloadurl =  ['text/json', a.download, a.href].join(':')
        e.initMouseEvent('click', true, false, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null)
        a.dispatchEvent(e)
    }
})(console)

//Common error reporting function
function reportError(){
	console.error('Oops, something went wrong :-(');
}

//Open the database
var dbResp = indexedDB.open('postman-app')
dbResp.onerror = reportError;
dbResp.onsuccess = function(){
	var db = dbResp.result;
	
	var responses = [];

	//Open cursor for saved responses
	var respReq = db.transaction(["history_responses"],"readonly").objectStore('history_responses').openCursor();
	respReq.onsuccess = function(event) {
		var respCur = event.target.result;
		if (respCur) {
			// respCur.value contains the current record being iterated through
			if(respCur.value.text && respCur.value.text.indexOf("#SUBSTRING#") !== -1) { // !!! To get unfiltered data comment this line!!!!
				responses.push(respCur.value);
			} // !!! To get unfiltered data comment this line!!!!
			respCur.continue();
		} else {
			console.save(JSON.stringify(responses, null, "\t"), 'postman-responses-export.json')
			console.info('Your responses have been exported to a file and downloaded to your computer.')
		}
	};
};
```

**Comment by valtlfelipe** [comment]
After upgrading to postman v8, it kind of requires me to use a workspace, but there is no option to move or export the history from the scratch pad to my workspace.

Any ideas how to proceed?

**Comment by zhiyong-ft** [comment]
> After upgrading to postman v8, it kind of requires me to use a workspace, but there is no option to move or export the history from the scratch pad to my workspace.
> 
> Any ideas how to proceed?

For those who struggle with transferring history to a new PC, try this: 

Copy and paste the following folder:

%USERPROFILE%\AppData\Roaming\Postman

I tested this with Postman V8.5.1 Windows App X64 version. 

Update on Nov. 15th 2024: this trick seems still work. This also can be a security hole, as all the sensitive information such as password are stored as plain text and can be easily recovered from a different computer.

**Comment by adriancuadrado** [comment]
# TLDR

Copy the Javascript code you want from [this gist](https://gist.github.com/adriancuadrado/84408b9b6d53e41858a2f1b9f20090d6) in the _Console_ tab of DevTools. [This comment](https://gist.github.com/adriancuadrado/84408b9b6d53e41858a2f1b9f20090d6?permalink_comment_id=4227724#gistcomment-4227724) explains how to do that.

Disclaimer: Previously I had the scripts copied in this comment, but I decided it would be better to put them in a separate place to maintain them and give people a way to add comments to it without overloading this thread.

**Comment by mrjoesm** [comment]
@adriancuadrado when I try this, I simply get a `{"requests":[],"responses":[]}` in the resultant `postman-history-export` file, have you ran into that?

**Comment by adriancuadrado** [comment]
@mrjoesm I ran the code and it worked for me. Make sure you select _Show DevTools (Current View)_ instead of _Show DevTools (Current Shell)_ under _View_ → _Developer_. Also, my Postman version is 9.22.2. Which is yours? You can check the version in the Gear icon in the top right corner → Release Notes.

Also I updated the comment to remove the code and include a reference to a Github gist where people can say what problems they have and make suggestions if they want. Please consider using that gist for anything related to the scripts I created instead of this thread. I don't want to overload this thread with more comments about my code.
